src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    public sealed partial class Semaphore
+    {
+        private const uint AccessRights = (uint)Interop.Kernel32.MAXIMUM_ALLOWED | Interop.Kernel32.SYNCHRONIZE | Interop.Kernel32.SEMAPHORE_MODIFY_STATE;
+
+        private Semaphore(SafeWaitHandle handle)
+        {
+            SafeWaitHandle = handle;
+        }
+
+        private void CreateSemaphoreCore(int initialCount, int maximumCount, string name, out bool createdNew)
+        {
+            Debug.Assert(initialCount >= 0);
+            Debug.Assert(maximumCount >= 1);
+            Debug.Assert(initialCount <= maximumCount);
+
+#if PLATFORM_UNIX
+            if (name != null)
+                throw new PlatformNotSupportedException(SR.PlatformNotSupported_NamedSynchronizationPrimitives);
+#endif
+            SafeWaitHandle myHandle = Interop.Kernel32.CreateSemaphoreEx(IntPtr.Zero, initialCount, maximumCount, name, 0, AccessRights);
+
+            int errorCode = Marshal.GetLastWin32Error();
```

Comment:
Why was this moved to not be guarded by the IsInvalid check?

Reply:
I see, it occurred in two different methods, in one before the if and in the other guarded by the if.  I suggest we move this to be inside the if block so that we don't call GetLastWin32Error unless we need to.

Reply:
CreateSemaphoreEx sets last error even when it returns valid handle. I think that it fine to always call GetLastWin32Error before the if block.

Reply:
But we don't do anything with it. Does GetLastWin32Error clear the error and that's why it's good to call regardless of whether the handle is valid?

Reply:
Oh, nevermind, we do do something with it.  Ignore me.